### PR TITLE
tee-supplicant: Fix crash when TA isn't found.

### LINKED ISSUE #374: `ta_path` in `set_ta_path()` doesn't add NULL to last element. It may cause segmentation fault.
Hi, 

When I upgrade optee os, optee client and optee test from 3.21.0 to 4.1.0, I met some trouble that xtest failed in `8002` and others test.
I traced `optee os` and `optee client` code and found a bug.

In optee client, function `set_ta_path` in `tee-supplicant/src/tee_supplicant.c` used to initialize the `ta_path`, which stores `load ta path`.
However, **`ta_path` doesn't add NULL to last element.**

https://github.com/OP-TEE/optee_client/blob/afbd31d9592e8919cf2b6883e33d80fcc32ccdc2/tee-supplicant/src/tee_supplicant.c#L701

Otherwise, In the function `TEECI_LoadSecureModule`,  it uses loop to traverse `ta_path and must be `NULL` to terminate.
If there isn't the `NULL` end of `ta_path`, it can't terminate loop and continue to move next pointer.
**Finally, it casued the  segmentation fault.**
https://github.com/OP-TEE/optee_client/blob/afbd31d9592e8919cf2b6883e33d80fcc32ccdc2/tee-supplicant/src/teec_ta_load.c#L160C1-L172C14
```
int TEECI_LoadSecureModule(const TEEC_UUID *destination, void *ta,
			   size_t *ta_size)
{
	int res = TA_BINARY_NOT_FOUND;
	char **path = NULL;

	for (path = ta_path; *path; path++) { // if path is NULL, it will terminate. 
		res = try_load_secure_module(*path, destination, ta, ta_size);
		if (res == TA_BINARY_FOUND)
			break;
	}
	return res;
}
```

**It caused the `tee-supplicant` segmentation fault.**
log is below
```
DBG [8586] TEES:try_load_secure_module:104: Attempt to load /lib/optee_armtz/17e5e280-d12e-11e4-a41a-0002a5d5c51b.ta
DBG [8586] TEES:try_load_secure_module:113: failed to open the ta /lib/optee_armtz/17e5e280-d12e-11e4-a41a-0002a5d5c51b.ta TA-file
DBG [8586] TEES:try_load_secure_module:104: Attempt to load /lib/optee_armtz/17e5e280-d12e-11e4-a41a0002a5d5c51b.ta
DBG [8586] TEES:try_load_secure_module:113: failed to open the ta /lib/optee_armtz/17e5e280-d12e-11e4-a41a0002a5d5c51b.ta TA-file
Segmentation fault
```

I have two solutions to solve this bug.
1. Replace `malloc` with `calloc` in `set_ta_path()`
https://github.com/OP-TEE/optee_client/blob/afbd31d9592e8919cf2b6883e33d80fcc32ccdc2/tee-supplicant/src/tee_supplicant.c#L725

2. Add NULL to `ta_path` last element after loop.
https://github.com/OP-TEE/optee_client/blob/afbd31d9592e8919cf2b6883e33d80fcc32ccdc2/tee-supplicant/src/tee_supplicant.c#L733
```
	while ((new_path = strtok_r(p, ":", &saveptr))) {
		if (!supplicant_params.ta_load_path) {
			char full_path[PATH_MAX] = { 0 };

			path_len = snprintf(full_path, PATH_MAX, "%s/%s", new_path,
					    supplicant_params.ta_dir);
			if (path_len < 0 || path_len >= PATH_MAX)
				goto err_path;

			ta_path[n++] = strdup(full_path);
		} else {
			path_len = strnlen(new_path, PATH_MAX);
			if (path_len == PATH_MAX)
				goto err_path;

			ta_path[n++] = strdup(new_path);
		}

		p = NULL;
	}
        ta_dir[n++] = NULL; // <---- add this line
```

If you need, I can launch PR to fix this bug.

Thanks !!

### PR DESCRIPTION
**Pull request by p870613**
In this PR , it fixes bug mentioned in #374.

fixes #374

### CODE REVIEW

**Review by jforissier — APPROVED** [review]
With my comment addressed:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: tee-supplicant/src/tee_supplicant.c [hunk 1]
```diff
@@ -750,7 +750,8 @@ static void set_ta_path(void)
 
 		p = NULL;
 	}
-
+	ta_path[n] = NULL;
+
```
[the remark this answers is not in the document]
Comment (p870613) [reply tee-supplicant/src/tee_supplicant.c]: done

File: tee-supplicant/src/tee_supplicant.c [hunk 2]
```diff
@@ -750,7 +750,7 @@ static void set_ta_path(void)
 
 		p = NULL;
 	}
-
+	ta_path[n] = NULL;
```
Comment (etienne-lms) [inline tee-supplicant/src/tee_supplicant.c:753]: I think a better fix would be to use `calloc()` instead of `malloc()` at line 725.

Comment (jforissier) [reply tee-supplicant/src/tee_supplicant.c:753]: I'm fine either ways

**Review by etienne-lms — COMMENTED** [review]
I've tested the _qemu_virt_ setup and I see that when a TA is not found, **tee-supplicant** does not crash. Do you have  a specific setup?

In my test, I run **tee-supplicant** manually (so that it's not restart from some init service), remove the TA related to regression test 1004, and test regression test 1004 and 2001. After 1004 failure, 2001 still succeeds:
```bash
(...)
buildroot login: root
sh> killall tee-supplicant
sh> tee-supplicant &
sh> xtest regression_1004
...
1 test case of which 0 failed
...
sh> rm /lib/optee_armtz/cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.ta
sh> xtest regression_1004
...
1 test case of which 1 failed
...
sh> xtest regression_2001
...
1 test case of which 0 failed
...
sh>

File: tee-supplicant/src/tee_supplicant.c (hunk 2 above)
Comment (etienne-lms) [reply tee-supplicant/src/tee_supplicant.c:753]: I think it's a much better habit to prefer `calloc()`  over `malloc()`. This is an example :)

Comment (p870613) [reply tee-supplicant/src/tee_supplicant.c:753]: If you think it's better way. I can modify this commit.

**Comment by p870613** [comment]
@etienne-lms 
I don't have any specific setting in our test environment.
I think it's a undefined behavior.
Crash point is in `TEECI_LoadSecureModule()` . In this function, it will traverse `ta_path` and it will terminate when it meet NULL pointer.
```
for (path = ta_path; *path; path++) {
		res = try_load_secure_module(*path, destination, ta, ta_size);
		if (res == TA_BINARY_FOUND)
			break;
	}
```
But, `malloc()` don't ensure that value is `0x00` and it keep the original value in memory for variable.

**Comment by etienne-lms** [comment]
I agree there is a bug and thank you for reporting and fixing it. By the way, could you add a reference to https://github.com/OP-TEE/optee_client/issues/374 in the commit message?
I doubt there'll be performance issues here using `calloc()`.

**Comment by p870613** [comment]
fixes #376

**Comment by p870613** [comment]
@etienne-lms
I modify the comment

Thanks !!

**Comment by jforissier** [comment]
@p870613  please use this commit subject and description:
```
tee-supplicant: fix potential crash when TA isn't found

set_ta_path() doesn't add a NULL pointer to the last element of ta_path,
which can cause a segmentation fault when a TA is not found. Use
calloc() instead of malloc() to make sure the ta_path array is always
NULL-terminated.

Signed-off-by: guan-gm.lin <guan-gm.lin@mediatek.com>
Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>
```

**Comment by p870613** [comment]
@jforissier 
Done

Thanks !!

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`
@p870613, could you update the commit message to add a `Link:` tag to the related Issue entry?
```
Link: https://github.com/OP-TEE/optee_client/issues/374

**Comment by p870613** [comment]
@etienne-lms 
Hi, 
Sorry, I am not sure if the commit message update like below or not.

```
tee-supplicant: fix potential crash when TA isn't found

set_ta_path() doesn't add a NULL pointer to the last element of ta_path,
which can cause a segmentation fault when a TA is not found. Use
calloc() instead of malloc() to make sure the ta_path array is always
NULL-terminated.

Link: https://github.com/OP-TEE/optee_client/issues/374

Signed-off-by: guan-gm.lin <guan-gm.lin@mediatek.com>
Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>
```

Thanks !!

**Comment by etienne-lms** [comment]
Remove the empty line between the `Link:` tag and your S-o-b tag, and it's all good.
Thanks.

**Comment by p870613** [comment]
Done

**Pull request merged**